Added unit tests for stats reporting service

Radesh-kumar opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #[fill_in_number_here].
2. This PR does the following: [Explain here what your PR does and why]

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Radesh-kumar committed:
created spec

Radesh-kumar committed:
initial

Radesh-kumar committed:
Merge remote-tracking branch 'upstream/develop' into stats_unit_tests

Radesh-kumar commented on core/templates/pages/exploration-player-page/services/stats-reporting.service.spec.ts:
```diff
@@ -0,0 +1,294 @@
+  });
+
+  it('should record exploration\'s stats when it is actually started', () => {
+    StatsReportingService.explorationActuallyStarted = false;
```
@nithusha21 your suggestion did the trick. But, both `StatsReportingService` and `statsReportingService` are not updating simultaneously in the tests.
If I place prechecks at the beginning of the test they are flaking (almost 90%  of the cases).
So in order to prevent that I've set the testing value to be null for each test.
(I feel that the ideal thing to do is change the static variable to public.)

Radesh-kumar replied:
PTAL @nithusha21

nithusha21 replied:
I think I like your suggestion to make it public. Do you know why it is static? In general, if the code feels off, feel free to make it correct.

Radesh-kumar replied:
I don't think we really need static here.
As it is not being used in hybrid state or in common anywhere else.
I'll try making them public and see if anything breaks.

Radesh-kumar committed:
Merge remote-tracking branch 'upstream/develop' into stats_unit_tests

nithusha21 reviewed (commented):
I think it looks good overall! Just need to sort out the static variable issue.

Radesh-kumar committed:
Merge remote-tracking branch 'upstream/develop' into stats_unit_tests

Radesh-kumar committed:
nits

Radesh-kumar committed:
coverage

Radesh-kumar commented:
PTAL @kevintab95 @nithusha21 @DubeySandeep 
It seems to be `stats-reporting-service` is not properly migrated previously.
That's why we have static variables inside the file.
I fixed the file and removed all static variables.

kevintab95 commented:
I think it was intentional to keep static variables in the StatsReportingService because we want to avoid potential inconsistencies due to multiple instantiation (similar to `PlatformFeatureService`). We use this service to store certain information in a player session (current/prev/next state names, explorationStarted, stateStopwatch etc.). If there are inconsistencies in these values in different instances of the service, we may be reporting incorrect stats to the backend. So please confirm this before actually removing the static variables. Thanks!

Radesh-kumar committed:
meged conflicts

Radesh-kumar commented:
> I think it was intentional to keep static variables in the StatsReportingService because we want to avoid potential inconsistencies due to multiple instantiation (similar to `PlatformFeatureService`). We use this service to store certain information in a player session (current/prev/next state names, explorationStarted, stateStopwatch etc.). If there are inconsistencies in these values in different instances of the service, we may be reporting incorrect stats to the backend. So please confirm this before actually removing the static variables. Thanks!

![Screenshot from 2021-06-28 16-47-59](https://user-images.githubusercontent.com/54100702/123628808-23c45a80-d831-11eb-8e3c-4ac52026c5e9.png)


I've checked the `stats reporting service` is being initialized only once.
Regarding to @srijanreddy98 comment [here](https://github.com/oppia/oppia/pull/10185/files#r473632233) static variables can be removed(in`PlatformFeatureService` too).
@srijanreddy98 Can you please confirm these changes are correct?

Radesh-kumar commented:
PTAL @srijanreddy98

DubeySandeep commented on core/templates/pages/exploration-player-page/services/stats-reporting.service.spec.ts:
```diff
@@ -0,0 +1,292 @@
+      .and.returnValue(Promise.resolve({}));
+    recordLeaveForRefresherExpSpy = spyOn(
+      statsReportingBackendApiService, 'recordLeaveForRefresherExpAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordStateHitSpy = spyOn(
+      statsReportingBackendApiService, 'recordStateHitAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordStateCompletedSpy = spyOn(
+      statsReportingBackendApiService, 'recordStateCompletedAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordExplorationCompletedSpy = spyOn(
+      statsReportingBackendApiService, 'recordExplorationCompletedAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordAnswerSubmittedSpy = spyOn(
+      statsReportingBackendApiService, 'recordAnswerSubmittedAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordMaybeLeaveEventSpy = spyOn(
+      statsReportingBackendApiService, 'recordMaybeLeaveEventAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordAnswerSubmitActionSpy = spyOn(
+      playthroughService, 'recordAnswerSubmitAction')
+      .and.returnValue(null);
+    spyOn(messengerService, 'sendMessage').and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerNewCard').and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerStartExploration')
+      .and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerFinishExploration')
+      .and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerCuratedLessonCompleted')
+      .and.returnValue(null);
+    spyOn(playthroughService, 'recordExplorationStartAction')
+      .and.returnValue(null);
+    spyOn(playthroughService, 'recordExplorationQuitAction')
+      .and.returnValue(null);
+    spyOn(playthroughService, 'storePlaythrough')
+      .and.returnValue(null);
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(true);
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getUrlParams')
+      .and.returnValue({classroom_url_fragment: 'classroom'});
```
Why are we creating so many spyOn in beforeEach block? It should be created in `it` block and only the required spyOn, right?

Radesh-kumar replied:
Most of the test cases use multiple spies.
In order to decrease the redundancy of code I've placed them in `beforeEach` block, that way it is easier to read code too.
I think both methods are identical.

DubeySandeep replied:
I think to make each test case readable we should write spies in `it` function where we are testing behavior based on the spied function!

Radesh-kumar replied:
done!

DubeySandeep reviewed (changes requested):
LGTM for the codeowner file, left one comment, PTAL!

srijanreddy98 commented:
> > I think it was intentional to keep static variables in the StatsReportingService because we want to avoid potential inconsistencies due to multiple instantiation (similar to `PlatformFeatureService`). We use this service to store certain information in a player session (current/prev/next state names, explorationStarted, stateStopwatch etc.). If there are inconsistencies in these values in different instances of the service, we may be reporting incorrect stats to the backend. So please confirm this before actually removing the static variables. Thanks!
> 
> ![Screenshot from 2021-06-28 16-47-59](https://user-images.githubusercontent.com/54100702/123628808-23c45a80-d831-11eb-8e3c-4ac52026c5e9.png)
> 
> I've checked the `stats reporting service` is being initialized only once.
> Regarding to @srijanreddy98 comment [here](https://github.com/oppia/oppia/pull/10185/files#r473632233) static variables can be removed(in`PlatformFeatureService` too).
> @srijanreddy98 Can you please confirm these changes are correct?

Yup, It can be removed.

Radesh-kumar commented:
PTAL @kevintab95

Radesh-kumar committed:
Merge remote-tracking branch 'upstream/develop' into stats_unit_tests

Radesh-kumar committed:
Merge remote-tracking branch 'upstream/develop' into stats_unit_tests

kevintab95 reviewed (approved):
Thanks for confirming the static variables stuff. LGTM!

Radesh-kumar commented:
PTAL @DubeySandeep

Radesh-kumar committed:
nits

Radesh-kumar committed:
nits

Radesh-kumar commented:
PTAL @DubeySandeep
